### PR TITLE
Issue 358 restyle dot graph

### PR DESCRIPTION
**Pull request by haim-kermany**
restyling the dot graph:
1. subgroup nodes into namespaces
2. represents cliques as nodes - to reduce number of edges
3. sorting edged names
4. restyling shapes and colors

updated expected results accordingly

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]
Great work.
Some comments (mostly minor changes).

File: nca/FWRules/DotGraph.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        """
+        add a node to the graph
+        param subgraph: subgraph name
+        param name: node name
+        param node_type: node type
+        param label: node label
+        """
+        label = [tok.strip() for tok in label if tok != '']
+        if subgraph not in self.subgraphs.keys():
+            self.subgraphs[subgraph] = self.Subgraph(subgraph)
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:54]: Currently, I put all the IPblocks under a subgraph calles 'external' (according to Ed Snible suggestion) but it is not related to this code.
We do have nodes not inside a subgraph - the clq nodes. for these nodes, the subgraph is an empty string.
Do you think we should to move the ipblocks outside the subgraph?

File: nca/FWRules/ConnectivityGraph.py [hunk 2]
```diff
@@ -48,48 +52,98 @@ def add_edges(self, connections):
...
-        if is_livesim:
-            return "coral4"
-        elif is_ip_block:
-            return "red2"
-        return "blue"
+
+        min_qlicue_size = 4
+
+        # replacing directed edges with not directed edges:
+        not_directed_edges = set([edge for edge in directed_edges if (edge[1], edge[0]) in directed_edges])
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/ConnectivityGraph.py:100]: done

File: nca/FWRules/ConnectivityGraph.py [hunk 3]
```diff
@@ -48,48 +52,98 @@ def add_edges(self, connections):
...
+        directed_edges = directed_edges - not_directed_edges
+        not_directed_edges = set([edge for edge in not_directed_edges if edge[1] < edge[0]])
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/ConnectivityGraph.py:102]: done

File: nca/FWRules/ConnectivityGraph.py [hunk 4]
```diff
@@ -48,48 +52,98 @@ def add_edges(self, connections):
...
+        graph = networkx.Graph()
+        graph.add_edges_from(not_directed_edges)
+        cliques = networkx.clique.find_cliques(graph)
+
+        cliques_nodes = []
+        cliques = sorted([sorted(clique) for clique in cliques])
+        for clique in cliques:
+            if len(clique) < min_qlicue_size:
+                continue
+            clq_namespaces = sorted(set([peer[1] for peer in clique]))
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/ConnectivityGraph.py:114]: done

File: nca/FWRules/ConnectivityGraph.py [hunk 5]
```diff
@@ -48,48 +52,98 @@ def add_edges(self, connections):
...
+                clq_namespace_peers = [peer for peer in clique if peer[1] == namespace_name]
+                if len(clq_namespace_peers) > 1:
+                    # creates a new clique node for the namespace
+                    namespace_clique_name = f'clique_{len(cliques_nodes)}'
+                    namespace_clique_node = (namespace_clique_name, namespace_name)
+                    cliques_nodes.append(namespace_clique_node)
+                    clique_namespaces_nodes.append(namespace_clique_node)
+
+                    # adds edges from the new node to the original clique nodes in the namespace
+                    not_directed_edges |= set([(namespace_clique_node, node) for node in clq_namespace_peers])
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/ConnectivityGraph.py:127]: done

File: nca/FWRules/ConnectivityGraph.py [hunk 6]
```diff
@@ -48,48 +52,98 @@ def add_edges(self, connections):
...
-            return peer.workload_name, False
-        return str(peer), False
+            name = peer.workload_name
+        else:
+            name = str(peer)
+        return name, peer_type, nc_name
 
-    @staticmethod
-    def _is_peer_livesim(peer):
+    def _creates_cliqued_graph(self, directed_edges):
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/ConnectivityGraph.py:75]: done

File: nca/FWRules/DotGraph.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    def add_node(self, subgraph, name, node_type, label):
+        """
+        add a node to the graph
+        param subgraph: subgraph name
+        param name: node name
+        param node_type: node type
+        param label: node label
+        """
+        label = [tok.strip() for tok in label if tok != '']
+        if subgraph not in self.subgraphs.keys():
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:53]: done

File: nca/FWRules/DotGraph.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+            return False
+        self.labels = list(self.labels)
+        self.labels.sort(reverse=True)
+
+        labels_tokens = {}
+        for label in self.labels:
+            # todo - we might need a better approach splitting the labels to tokens
+            # we should revisit this code after reformatting connections labels
+            labels_tokens[label] = re.findall(r"[\w']+", label)
+        first_tokens = set([t[0] for t in labels_tokens.values()])
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:176]: done

File: nca/FWRules/DotGraph.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        for first_token in first_tokens:
+            token_labels = [label for label in labels_tokens.keys() if labels_tokens[label][0] == first_token]
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:178]: done

File: nca/FWRules/DotGraph.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            self.name = name
+            self.node_type = node_type
+            self.label = label
+
+    class Edge:
+        def __init__(self, src, dst, label, is_dir):
+            self.src = src
+            self.dst = dst
+            self.label = label
+            self.is_dir = is_dir
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:29]: done

File: nca/FWRules/DotGraph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply nca/FWRules/DotGraph.py:54]: I don't think grouping all IP-Blocks in a box makes much sense. It's not as if Google's IPs have something to do with Amazon's IPs. @adisos , what do you think?

File: nca/FWRules/DotGraph.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        """
+        if not self.labels:
+            return False
+        if len(max(self.labels, key=len)) <= 11:
+            self.labels_dict = {label: label for label in self.labels}
+            return False
+        self.labels = list(self.labels)
+        self.labels.sort(reverse=True)
+
+        labels_tokens = {}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:171]: correct, the sort in not in the right place - fixed and documented

File: nca/FWRules/DotGraph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adisos) [reply nca/FWRules/DotGraph.py:54]: Is it related to grouping if they have the exact same ingress/egress connections?

[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:54]: no, it is related to that all nodes of the same namespace are in the same subgraph.

[the remark this answers is not in the document]
Comment (adisos) [reply nca/FWRules/DotGraph.py:54]: so I think I agree with @zivnevo , and would not group all IP blocks

File: nca/FWRules/DotGraph.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+    class Subgraph:
+        def __init__(self, name):
+            self.name = name
+            self.nodes = []
+
+    class Node:
+        def __init__(self, name, node_type, label):
+            self.name = name
+            self.node_type = node_type
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py]: done

File: nca/FWRules/DotGraph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply nca/FWRules/DotGraph.py:62]: so moved the ip-blocks from subgraph - done

**Review by zivnevo — APPROVED** [review]

**Pull request merged**